code/SecurityAdmin.php

```diff
@@ -3,314 +3,145 @@
-
-        return $this->renderWith('BlankPage', [
-            'Form' => $this->MemberImportForm()->forTemplate(),
-            'Content' => ' '
-        ]);
+        return $models;
     }
 
     /**
-     * @see SecurityAdmin_MemberImportForm
-     *
-     * @return Form
+     * @return Form|false
      */
-    public function MemberImportForm()
+    public function ImportForm()
     {
-        if (!Permission::check('ADMIN')) {
-            return null;
+        $form = parent::ImportForm();
+        if (!$form) {
+            return $form;
         }
-
-        /** @var Group $group */
-        $group = $this->currentPage();
-        $form = new MemberImportForm($this, __FUNCTION__);
-        $form->setGroup($group);
-
-        return $form;
-    }
-
-    public function groupimport()
-    {
-        Requirements::clear();
-        Requirements::javascript('silverstripe/admin: client/dist/js/vendor.js');
-        Requirements::javascript('silverstripe/admin: client/dist/js/MemberImportForm.js');
-        Requirements::css('silverstripe/admin: client/dist/styles/bundle.css');
-
-        return $this->renderWith('BlankPage', [
-            'Content' => ' ',
-            'Form' => $this->GroupImportForm()->forTemplate()
+        Requirements::javascript('silverstripe/admin:client/dist/js/MemberImportForm.js');
+        $form->Fields()->removeByName('EmptyBeforeImport');
+        $name = match ($this->modelClass) {
+            Member::class => 'SpecForMember',
+            Group::class => 'SpecForGroup',
+        };
+        $form->Fields()->removeByName($name);
+        $class = match ($this->modelClass) {
+            Member::class => MemberCsvBulkLoader::class,
+            Group::class => GroupCsvBulkLoader::class,
+        };
+        $importSpec = $class::create()->getImportSpec();
+        $columns = implode(', ', array_keys($importSpec['fields'] ?? []));
+        $helpHtml = implode("\n", [
+            _t(
+                __CLASS__ . '.Help1',
+                '<p>Import users in <em>CSV format</em> (comma-separated values).'
+                . ' <small><a href="#" class="toggle-advanced">Show advanced usage</a></small></p>'
+            ),
+            _t(
+                __CLASS__ . '.Help2',
+                '<div class="advanced">'
+                . '<h4>Advanced usage</h4>'
+                . '<ul>'
+                . '<li>Allowed columns: <em>{columns}</em></li>'
+                . '<li>Existing users are matched by their unique <em>Code</em> property, and updated with any new values from '
+                . 'the imported file.</li>'
+                . '<li>Groups can be assigned by the <em>Groups</em> column. Groups are identified by their <em>Code</em> property, '
+                . 'multiple groups can be separated by comma. Existing group memberships are not cleared.</li>'
+                . '</ul>'
+                . '</div>',
+                ['columns' => $columns]
+            )
```

Comment:
The allowed columns is already in the `ModelAdmin` import form - it feels like overkill to be replacing that with a less-standard way of displaying that same information.

From what I can tell, the only thing this _adds_ on top of what `ModelAdmin` has in the form already is the second two bullet points there which only apply to `Member` - and the first point I think isn't correct so we should probably remove that. `Member` has no `Code` property.
There's also some information missing from the `GroupImportForm` which applies only to groups.

I think it would be cleaner to add that information like this:
```php
public function ImportForm()
{
    $form = parent::ImportForm();
    if (!$form) {
        return $form;
    }
    $form->Fields()->removeByName('EmptyBeforeImport');
    $extraInfo = match ($this->modelClass) {
        Member::class => '<p>Groups can be assigned by the <em>Groups</em> column. Groups are identified by their '
            . '<em>Code</em> property, multiple groups can be separated by comma. Existing group memberships are '
            . 'not cleared.</p>',
        Group::class => '<ul><li>Existing groups are matched by their unique <em>Code</em> value, and updated with any new values from the '
        . 'imported file</li>'
        . '<li>Group hierarchies can be created by using a <em>ParentCode</em> column.</li>'
        . '<li>Permission codes can be assigned by the <em>PermissionCode</em> column. Existing permission codes are not '
        . 'cleared.</li></ul>'
    };
    if ($extraInfo) {
        $form->Fields()->insertBefore('_CsvFile', LiteralField::create('ExtraInfo', $extraInfo));
    }
    return $form;
}
```

This way `ModelAdmin` is still in charge of saying what columns are valid, and it does so in a standard format that people will be used to from everywhere else they can import CSV content. The columns mentioned are the same as what you're adding here.
It also means we can get rid of that extra javascript which is essentially just a duplication of the javascript `ModelAdmin` already has but with different selectors.


An alternative, if you'd specifically like this information to be inside the "Show specification" expandable area, is to add a `$ExtraDetails` section into the `ModelAdmin_ImportSpec.ss` template and introduce a new `protected function getExtraImportDetails()` method, which `ModelAdmin` can call to add any extra details into the template. If there are no extra details then it'll just render as it always has, but if there _are_ extra details, it can add a new section inside that expandable area to display it. That feels a lot cleaner and it also has the benefit of providing that functionality to custom `ModelAdmin` in people's projects.

Reply:
I've also just read up on `match` expressions - we need to add a `default` here in case someone adds import functionality to permission codes or it will raise a `E_FATAL_ERROR`

Reply:
Updated